extensions/amp-share-tracking/0.1/amp-share-tracking.js

```diff
@@ -15,43 +15,116 @@
+import {xhrFor} from '../../../src/xhr';
+import {viewerFor} from '../../../src/viewer';
+import {dev, user} from '../../../src/log';
 
 /** @private @const {string} */
 const TAG = 'amp-share-tracking';
 
-class AmpShareTracking extends AMP.BaseElement {
+export class AmpShareTracking extends AMP.BaseElement {
+  /**
+    * @return {boolean}
+    * @private
+    */
+  isExperimentOn_() {
+    return isExperimentOn(this.win, TAG);
+  }
+
+  /** @override */
+  isLayoutSupported(unusedLayout) {
+    return true;
+  }
 
   /** @override */
   buildCallback() {
-    this.isExperimentOn_ = isExperimentOn(this.win, TAG);
-    if (!this.isExperimentOn_) {
+    if (!this.isExperimentOn_()) {
       dev.warn(TAG, `TAG ${TAG} disabled`);
       return;
     }
+
+    /** @private {string} */
+    this.vendorHref_ = this.element.getAttribute('data-href');
+    dev.fine(TAG, 'vendorHref_: ', this.vendorHref_);
+
+    /** @private {!Promise<!Object>} */
+    this.shareTrackingFragments_ = Promise.all([
+      this.getIncomingFragment_(),
+      this.getOutgoingFragment_(this.vendorHref_)]).then(results => {
+        dev.fine(TAG, 'incomingFragment: ', results[0]);
+        dev.fine(TAG, 'outgoingFragment: ', results[1]);
+        return {
+          incomingFragment: results[0],
+          outgoingFragment: results[1],
+        };
+      });
   }
-}
 
-/**
- * ShareTrackingService processes the incoming and outcoming url fragment
- */
-export class ShareTrackingService {
-  constructor(window) {
-    this.win = window;
+  /**
+   * Get the incoming share-tracking fragment from the viewer
+   * @return {!Promise<string|undefined>}
+   * @private
+   */
+  getIncomingFragment_() {
+    return viewerFor(this.win).getFragment().then(fragment => {
+      if (!fragment || fragment.indexOf('.') != 0) {
+        return;
+      }
+      const endIndex = fragment.indexOf('&');
+      if (endIndex != -1) {
+        return fragment.substr(1, (endIndex - 1));
+      } else {
+        return fragment.substr(1);
+      }
+    });
   }
-}
 
-/**
- * @param {!Window} window
- * @return {!ShareTrackingService}
- */
-export function installShareTrackingService(window) {
-  return getService(window, 'shareTrackingService', () => {
-    return new ShareTrackingService(window);
-  });
-}
+  /**
+   * Get an outgoing share-tracking fragment
+   * @param {string=} opt_vendorUrl
+   * @return {!Promise<string|undefined>}
+   * @private
+   */
+  getOutgoingFragment_(opt_vendorUrl) {
+    if (opt_vendorUrl) {
+      return this.getOutgoingFragmentFromVendor_(opt_vendorUrl);
+    }
+    return this.getOutgoingRandomFragment_();
+  }
 
-installShareTrackingService(AMP.win);
+  /**
+   * Get an outgoing share-tracking fragment from vendor
+   * by issueing a post request to the url the vendor provided
+   * @param {string} vendorUrl
+   * @return {!Promise<string|undefined>}
+   * @private
+   */
+  getOutgoingFragmentFromVendor_(vendorUrl) {
+    const postReq = {
```

Comment:
We usually require `requireAmpResponseSourceOrigin: true` along with `credentials: include`. Is there a reason why this is not the case here? (it possible that this is completely fine, but i'd like to make sure)